plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1219 @@
+    * Decorator does NOT perform any unit conversions, look to
+      :func:`~plasmapy.utils.decorators.validate_quantities` if that functionality is
+      desired.
+    * If it is desired that `None` values do not raise errors or warnings, then
+      include `None` in the list of units or as a default value for the function
+      argument.
+    * If units are not specified in `checks`, then the decorator will attempt
+      to identify desired units by examining the function annotations.
+    * Full functionality is defined by the class :class:`CheckUnits`.
+
+    Examples
+    --------
+    Define units with decorator parameters::
+
+        import astropy.units as u
+        from plasmapy.utils.decorators import check_units
+
+        @check_units(arg1={'units': u.cm},
+                     arg2=u.cm,
+                     checks_on_return=[u.cm, u.km])
+        def foo(arg1, arg2):
+            return arg1 + arg2
+
+        # or on a method
+        class Foo:
+            @check_units(arg1={'units': u.cm},
+                         arg2=u.cm,
+                         checks_on_return=[u.cm, u.km])
+            def bar(self, arg1, arg2):
+                return arg1 + arg2
+
+    Define units with function annotations::
+
+        import astropy.units as u
+        from plasmapy.utils.decorators import check_units
+
+        @check_units
+        def foo(arg1: u.cm, arg2: u.cm) -> u.cm:
+            return arg1 + arg2
+
+        # or on a method
+        class Foo:
+            @check_units
+            def bar(self, arg1: u.cm, arg2: u.cm) -> u.cm:
+                return arg1 + arg2
+
+    Allow `None` values to pass::
+
+        import astropy.units as u
+        from plasmapy.utils.decorators import check_units
+
+        @check_units(checks_on_return=[u.cm, None])
+        def foo(arg1: u.cm = None):
+            return arg1
+
+    Allow return values to have equivalent units::
+
+        import astropy.units as u
+        from plasmapy.utils.decorators import check_units
+
+        @check_units(arg1={'units': u.cm},
+                     checks_on_return={'units': u.km,
+                                       'pass_equivalent_units': True})
+        def foo(arg1):
+            return arg1
+
+    Allow equivalent units to pass with specified equivalencies::
+
+        import astropy.units as u
+        from plasmapy.utils.decorators import check_units
+
+        @check_units(arg1={'units': u.K,
+                           'equivalencies': u.temperature(),
+                           'pass_equivalent_units': True})
+        def foo(arg1):
+            return arg1
+
+    .. _astropy equivalencies:
+        https://docs.astropy.org/en/stable/units/equivalencies.html
+    """
+    if checks_on_return is not None:
+        checks['checks_on_return'] = checks_on_return
+
+    if func is not None:
+        return CheckUnits(**checks)(func)
+    else:
+        return CheckUnits(**checks)
```

Comment:
E L E G A N T

I also like how the docstring for this is 20x longer. :laughing: 

Seriously, though, well refactored :)

Reply:
Thank you! :)